Draft for Spring Boot 3 WAR post

scottkurz opened the pull request:
Fixes https://github.com/OpenLiberty/blogs/issues/3539.

I realize there are quite a few things not polished yet.

But I was hoping to start getting a review with the general outline of the content.   (This is more than an outline, it's the content itself... but what I'm saying is that the question of what we include could still be up for debate.) 

So I'll just cc: 
@tjwatson @cbridgha @yeekangc @lauracowen @hlhoots 
and wait for a PR review until it's more polished.

Also we didn't pin down the key question of does this replace the earlier [article](https://openliberty.io/blog/2023/06/15/running-spring-boot-3.html) more "inline" or do we do more of delete the old and create the new?

Also, I'm not sure how much detail is needed for the minor/micro release support... e.g. 3.1.x vs 3.2.x , vs 3.2.3 vs 3.2.4 etc.   I'm confused about that.  
E.g. in  24.0.0.1 (https://github.com/OpenLiberty/open-liberty/pull/27276)  we say we have support for 3.2 thinning

Maybe this needs a doc update too?  Opened doc issue:  https://github.com/OpenLiberty/docs/issues/7284

scottkurz commented:
Maybe I should add into the blog post, for completeness,  that if you choose a JAR packaging type your only option is to use the optimized deployment with `<feature>springBoot-3.0</feature>`

scottkurz commented:
I did get one internal comment suggesting showing more of the Java and server.xml.   IMO, this sort of would amount to having a new or updated guide.  CC'ing @gkwan-ibm as feedback/input.

gkwan-ibm commented:
hi @scottkurz 

would you explain more what need to be updated in the existing/new guide? Or even better if you can open an issue on the guide-spring-boot or guides-common repo. Thanks

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+As of Liberty version 23.0.0.9 there are now two ways to deploy Spring Boot WAR files on Liberty.
```
I know this hasn't been addressed in any of our other blog posts/guide about SB support yet either, but I think this sentence should be followed by answering the question of "why would I do that [deploy a Spring Boot WAR on Liberty]?". Without it, yes, the post is providing some helpful documentation for people who need it, but a blog post should provide more than that and it would be good to also convince readers that it's a valuable thing to do for relevant use cases.

lauracowen replied:
eg to benefit from InstantOn for a faster startup whilst maintaining throughput and not having to modify the app?

scottkurz replied:
The answer to "why" might depend on your baseline alternative

 * not going to get into why choose WAR but assume they know.  I even feel like "..to take advantage of the Spring and Spring Boot libraries " is too obvious to mention but leaving it for now as a compromise
 * follow-up.. this good to reference: 

That answers why SB on Liberty.

As far as why "optimized deployment"?  That's addressed with:

>  This deployment adds some additional configuration options, and provides advantages for efficiently building Docker container images

OTOH, InstantOn is supposed to work with both deployments.

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+
+As of Liberty version 23.0.0.9 there are now two ways to deploy Spring Boot WAR files on Liberty.
+
+While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, it can also be deployed using an optimized deployment for Spring Boot applications. This deployment adds some additional configuration options, and provides advantages for efficiently building Docker container images.
```
```suggestion
While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, the WAR can be optimized deployment for Spring Boot applications. The optimization adds some additional configuration options, and provides advantages for efficiently building Docker container images.
```

lauracowen replied:
Also, why would you ever use a non-optimized approach? It sounds, um, sub-optimal...

scottkurz replied:
* This part "..the WAR can be optimized deployment .." read a bit awkward so I took your suggestion and tweaked a bit


"Also, why would you ever use a non-optimized approach? It sounds, um, sub-optimal..."

Because:
 * you need extra pom.xml config
 * you need special server.xml config
 * you can't use dev mode

and because you might want to deploy like any WAR.


I added:
"using some extra pom.xml and server.xml configuration."

to maybe thread the needle here or feel free to suggest another wording.

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+
+While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, it can also be deployed using an optimized deployment for Spring Boot applications. This deployment adds some additional configuration options, and provides advantages for efficiently building Docker container images.
+
+In this blog post we will step through a simple flow starting from the Spring Initializr.  We will deploy this to Liberty via a standard WAR deployment without any Liberty modifications to the app.  
```
```suggestion
In this blog post, we create a simple app by using the Spring Initializr, and then deploy the app as a WAR to Liberty without any modifications.
```

scottkurz replied:
Trying to emphasize the "typical deployment" contrasted with the later optimized, so added to your wording"

"..., using the same deployment as for any WAR"

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+
+In this blog post we will step through a simple flow starting from the Spring Initializr.  We will deploy this to Liberty via a standard WAR deployment without any Liberty modifications to the app.  
+
+From there we'll show how to deploy using the optimized Spring Boot deployment by adding the liberty-maven-plugin and some minimal config to the pom.xml, and simple server.xml configuration using the `springBoot-3.0` feature.   We will then be able to build an efficiently-layered Docker image.
```
```suggestion
We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and some minimal config to the `pom.xml`, and add a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
```

scottkurz replied:
Mostly took your change ..but left the words "the optimized deployment".  This deployment is a direction that drags in several ramifications.  When you go this route, you have to have special pom.xml, server.xml config.. 
It even affects a bit of the programming model.

It's not just something you do in the final step of deployment.  It's a choice you have to make.

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+
+From there we'll show how to deploy using the optimized Spring Boot deployment by adding the liberty-maven-plugin and some minimal config to the pom.xml, and simple server.xml configuration using the `springBoot-3.0` feature.   We will then be able to build an efficiently-layered Docker image.
+
+Much of the Liberty documentation relating to Spring Boot focuses on the optimized deployment associated with the Spring Boot features, and so we spend some extra time in this post explaining both options.
```
Can you add a link to that Liberty doc?

lauracowen replied:
And also say why the Liberty doc focuses mainly on the optimized approach and, conversely, why it's worth covering both in this post? Otherwise it sounds like you're criticising the Liberty doc (instead of just fixing the doc), and it's not clear why the Liberty doc would do one thing and you'd do something else in the blog post?

scottkurz replied:
Done.   Also opened https://github.com/OpenLiberty/docs/issues/7296 for doc follow-up... and had opened https://github.com/OpenLiberty/docs/issues/7284 for tracking the major.minor version support. (edited)

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+== Prerequisites
+- Java 21 (sample could be easily reworked with Java 17)
+
+== Deploy as standard WAR
```
I know this is an initial draft, but just so we don't forget: expand the sentences into full sentences throughout.

scottkurz replied:
Done

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
++
+The app is bootstrapped by the web container via the ServletInitializer, extending `SpringBootServletInitializer`.  We go out of our way to run this example with only default Liberty and Maven configuration even though it'd be more typical to maintain a server.xml configuration for your app.
+
+== Use Liberty Spring Boot feature deployment
```
I don't think I understand what this heading means and how it relates back to what you discussed in the intro. Does this map to the "optimized deployment" approach?

scottkurz replied:
Reworded.  Yes, exactly.  I was hoping the reader would make the leap to equate the two ("optimized" == "Spring Boot feature") but the fact that you asked shows that was wrong...should be more explicit

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+
+We're going to reuse the app from the previous example.
+
+1. Add liberty-maven-plugin with Spring Boot deployment config to pom.xml:
```
ok, but what does this do for the app?

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+    </plugin>
+----
++
+2. Create your Liberty server.xml by copying from the Spring Boot template
```
And what does this do for the app?

scottkurz replied:
Did my extra sentence address your concern?

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+cp target/liberty/wlp/usr/servers/defaultServer/server.xml src/main/liberty/config/server.xml
+----
++
+3. Add application arguments to your Spring Boot application configuration in Liberty server.xml
```
What do these do for the app?

scottkurz replied:
Did my extra sentence address your concern?

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
++
+5. Run
++
+Though we won't benefit from using dev mode, we can still use the "lifecycle" built into the 'run' goal.  We do need to do a 'package' first to get the Spring Boot "repackage" packaging.
```
Why not? (I know dev mode doesn't support SB yet but I think that needs to be said explicitly (somewhere) otherwise it's just a bit mysterious.)

scottkurz replied:
Explained

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+[INFO] [AUDIT   ] CWWKT0016I: Web application available (default_host): http://localhost:9080/testpath/
+----
++
+7. Test the application by visiting the following URL in a browser: http://localhost:9080/testpath/
```
What should you see at that URL?

scottkurz replied:
TODO - I owe you an answer.  Will set up and run

scottkurz replied:
DONE

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
+./mvnw package liberty:run
+----
++
+6.  Observe the output:
```
What if the reader/user doesn't see that output? What's most likely gone wrong?

scottkurz replied:
Good question.  I don't have an answer though.  If I think of one I'll add it

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
++
+The server is configured with the https://openliberty.io/docs/latest/reference/feature/springBoot-3.0.html[springBoot-3.0] feature. The liberty-maven-plugin has been added to the pom.xml along with special configuration to use the optimized Spring Boot deployment.  The app is packaged as an "executable WAR" by running the `spring-boot:repackage` goal in the `package` phase, and the app is bootstr ...
++
+If you're wondering, yes, the "executable WAR" binary package here could still be deployed as a standard WAR (though a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment).
```
```suggestion
If you're wondering, yes, while a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment, that executable WAR could still be deployed as a standard WAR.
```

scottkurz replied:
Took

lauracowen commented on posts/2024-xx-xx-spring-boot-3.adoc:
```diff
@@ -0,0 +1,223 @@
++
+If you're wondering, yes, the "executable WAR" binary package here could still be deployed as a standard WAR (though a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment).
+
+== Build container image with efficient layering
```
Should this be an H3, rather than H2? It seems to relate only to the optimized instructions?

scottkurz replied:
Good catch, you're thorough :)

lauracowen reviewed (changes requested):
I think I'd need to @yeekangc to take a look when he's back regarding the overall post and what should/shouldn't be included, I've made some comments about needing more explanation/context (mainly). I know Chuck's post is quite sparse but I think similar comments should apply there too really.

If you can run a build on the draft server, we can see how it looks for length but it doesn't seem too long. My feeling is that the level of detail for the instructions is about right, there just needs to be more explanation (why) of what the user is being asked to do.

scottkurz commented:
> hi @scottkurz
> 
> would you explain more what need to be updated in the existing/new guide? Or even better if you can open an issue on the guide-spring-boot or guides-common repo. Thanks

I don't feel strongly enough to push this forward for now, so I'll just stop with this previous comment.   It's not that I think there's one clear next step so it would depend what we wanted to show.

scottkurz commented:
@lauracowen, thanks for the review.

One overall comment:  as far as the word choice "optimized".  I'm open to other suggestions. I'm far from a Spring Boot expert.  It seemed a decent mix of vague and helpful to at least use in the blog post.   What I really don't know is if it's good enough to promote to the doc.

I think it's better than "..using the Spring Boot feature" which would beg the question of what the feature provides.

Opened https://github.com/OpenLiberty/docs/issues/7296 for this... and had opened https://github.com/OpenLiberty/docs/issues/7284 for tracking the major.minor version support. (edited) 

Pushed a commit and addressed your comments 1-at-a-time. 

 Now that I'm done. I think if I'd addressed them holistically I'd have done better, but I'm just going to push that out there for further review

hlhoots commented:
Looks good to me.  Do we want a Reference in that last section to point to the Spring Boot OL guide: https://openliberty.io/guides/spring-boot.html?

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+As of Liberty version 23.0.0.9 there are now two ways to deploy Spring Boot WAR files on Liberty.  This allows developers to take advantage of the Spring and Spring Boot libraries while leveraging the performance enhancements Liberty has built especially for Spring Boot applications, addressing https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance] throughput  ...
```
```suggestion
In Liberty version 23.0.0.9 and later, you can get enhancements for Spring Boot applications that address https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance], throughput, memory footprint, and rapid application startup through https://openliberty.io/blog/2023/09/26/spring-boot-3-instant-on.html[InstantOn].
```

dmuelle replied:
We call out InstantOn here, but it isn't otherwise mentioned or explained in this post. We should either add some content about it or remove this reference

scottkurz replied:
OK, removed refs to InstantOn.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+As of Liberty version 23.0.0.9 there are now two ways to deploy Spring Boot WAR files on Liberty.  This allows developers to take advantage of the Spring and Spring Boot libraries while leveraging the performance enhancements Liberty has built especially for Spring Boot applications, addressing https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance] throughput  ...
+
+While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, the WAR can be optimized via a special deployment for Spring Boot applications. The optimization adds some additional configuration options, and provides advantages for efficiently building Docker container images, using some extra pom.xml and server.xml configuration.
```
```suggestion
Although you can deploy a Spring Boot WAR like any standard Jakarta EE WAR, with Liberty, you can also optimize it with a special deployment for Spring Boot applications. The optimization adds some configuration options and provides advantages for efficiently building Docker container images by using some extra `pom.xml` and `server.xml` file configuration.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+While a Spring Boot WAR can be deployed like any standard Jakarta EE WAR, the WAR can be optimized via a special deployment for Spring Boot applications. The optimization adds some additional configuration options, and provides advantages for efficiently building Docker container images, using some extra pom.xml and server.xml configuration.
+
+In this blog post, we create a simple app by using the Spring Initializr, and then deploy the app as a WAR to Liberty without any modifications, using the same deployment as for any WAR.
```
```suggestion
In this blog post, we'll first create a simple app by using the link:https://start.spring.io/[Spring Initializr], and then deploy it as a WAR to Liberty without any modifications, using the same deployment as for any WAR.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+In this blog post, we create a simple app by using the Spring Initializr, and then deploy the app as a WAR to Liberty without any modifications, using the same deployment as for any WAR.
+
+We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and some minimal config to the `pom.xml`, and add a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
```
```suggestion
We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and some minimal config to the `pom.xml`, and add a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
+
+Though the Spring Boot https://openliberty.io/docs/latest/deploy-spring-boot.html[documentation] in Open Liberty focuses mostly on the optimized deployment associated with the Spring Boot features, there are a couple advantages with using a traditional WAR deployment, which can use the same `server.xml` and `pom.xml` configuration, and maybe the same Docker build, as your non-Spring Boot WARs. So ...
```
I would move this paragraph up to be after the one at line 22 to make the case for including two use cases before you detail what those use cases are in the following paragraphs.

scottkurz replied:
Reworked.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+== Deploy as standard WAR
+
+1. Create project from Spring Initializr
```
```suggestion
1. Create a project project from your browser using link:https://start.spring.io/#!type=maven-project&language=java&platformVersion=3.2.4&packaging=war&jvmVersion=21&groupId=demo&artifactId=spring-liberty&name=spring-liberty&description=Demo%20project%20for%20Spring%20Boot&packageName=demo&dependencies=web[this Spring Initializr configuration].
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+1. Create project from Spring Initializr
++
+From browser:  https://start.spring.io/#!type=maven-project&language=java&platformVersion=3.2.4&packaging=war&jvmVersion=21&groupId=demo&artifactId=spring-liberty&name=spring-liberty&description=Demo%20project%20for%20Spring%20Boot&packageName=demo&dependencies=web
```
```suggestion
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+From browser:  https://start.spring.io/#!type=maven-project&language=java&platformVersion=3.2.4&packaging=war&jvmVersion=21&groupId=demo&artifactId=spring-liberty&name=spring-liberty&description=Demo%20project%20for%20Spring%20Boot&packageName=demo&dependencies=web
++
+2. Add web controller `vi src/main/java/demo/HelloController.java`
```
```suggestion
2. Add the following `vi src/main/java/demo/HelloController.java` web controller file.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+}
+----
++
+3. Run with standard WAR deployment with default configuration
```
```suggestion
3. From the command line, run a standard WAR deployment with the default configuration.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+  Hello from Spring Boot in Open Liberty!
++
+5. Stop the server (with  `<Ctrl>+C` or `q,<Enter>` to quit dev mode).
```
```suggestion
5. Stop the server by pressing `Ctrl`+`C` or  by typing `q` and pressing `Enter` to quit dev mode.
```

but does the mvn command in step 3 start server in dev mode?

scottkurz replied:
Added a mention of dev mode in 3.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+5. Stop the server (with  `<Ctrl>+C` or `q,<Enter>` to quit dev mode).
++
+6. Recap
```
```suggestion
=== Results
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+6. Recap
++
+The app is bootstrapped by the web container via the ServletInitializer, extending `SpringBootServletInitializer`.  We go out of our way to run this example with only default Liberty and Maven configuration even though it'd be more typical to maintain a server.xml configuration for your app.
```
the results should make the benefit of running a standard WAR more clear. These results focus on config details as opposed to value add

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+== Use optimized Liberty Spring Boot deployment
+
+We're going to reuse the app from the previous example.   By adding some extra configuration in the `pom.xml` and `server.xml` we end up being able to build an efficiently-layered Docker image, and also to parameterize the app with application arguments.
```
```suggestion
This example uses the same Spring app that is created in step 1 of the previous example. By adding some extra configuration in the `pom.xml` and `server.xml`, you can build an efficiently-layered Docker image and also to parameterize the app with application arguments.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+We're going to reuse the app from the previous example.   By adding some extra configuration in the `pom.xml` and `server.xml` we end up being able to build an efficiently-layered Docker image, and also to parameterize the app with application arguments.
+
+1. Add liberty-maven-plugin with Spring Boot deployment config to pom.xml:
```
```suggestion
1. Add liberty-maven-plugin with Spring Boot deployment config to the `pom.xml` file:
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+    </plugin>
+----
++
+2. Create your Liberty server.xml by copying from the Spring Boot template
```
```suggestion
2. Create your Liberty `server.xml` file by copying from the Spring Boot template
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+cp target/liberty/wlp/usr/servers/defaultServer/server.xml src/main/liberty/config/server.xml
+----
++
+3. Add application arguments to your Spring Boot application configuration in Liberty server.xml
```
```suggestion
3. Add application arguments to your Spring Boot application configuration in the Liberty `server.xml` file
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+ </springBootApplication>
+----
++
+4. Add println to echo args in output, `vi ./src/main/java/demo/SpringLibertyApplication.java`
```
```suggestion
4. Add `println` to echo arguments in the `vi ./src/main/java/demo/SpringLibertyApplication.java` output.
```

scottkurz replied:
See what you think of this.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+5. Run
++
+Though the Liberty Maven Plugin doesn't currently have support for using dev mode (or "loose applications") when using the optimized deployment, we can still use the "lifecycle" built into the 'run' goal.  We do need to do a 'package' first to get the Spring Boot "repackage" packaging.
```
```suggestion
Although the Liberty Maven Plugin doesn't currently support dev mode when you use the optimized deployment, you can still use the life cycle built into the 'run' goal.  We do need to do a 'package' first to get the Spring Boot "repackage" packaging.
```

scottkurz replied:
Took this but kept "lifecycle" as one word since it's meant to track the Maven term "lifecycle".

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+  Hello from Spring Boot in Open Liberty!
++
+8. When you finish testing the application, stop the server by typing `<Ctrl>+C`.
```
```suggestion
8. When you finish testing the application, stop the server by pressing `Ctrl`+`C`.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+8. When you finish testing the application, stop the server by typing `<Ctrl>+C`.
++
+9. Recap
```
```suggestion
=== Results
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
++
+9. Recap
++
+The server is configured with the https://openliberty.io/docs/latest/reference/feature/springBoot-3.0.html[springBoot-3.0] feature. The liberty-maven-plugin has been added to the pom.xml along with special configuration to use the optimized Spring Boot deployment.  The app is packaged as an "executable WAR" by running the `spring-boot:repackage` goal in the `package` phase, and the app is bootstr ...
```
```suggestion
The server is configured with the https://openliberty.io/docs/latest/reference/feature/springBoot-3.0.html[springBoot-3.0] feature. The `liberty-maven-plugin` is added to the `pom.xml`, along with special configuration to use the optimized Spring Boot deployment.  The app is packaged as an executable WAR by running the `spring-boot:repackage` goal in the `package` phase, and the app is bootstrapped by its `main()` method in `SpringLibertyApplication` passing in application arguments defined in `server.xml`.
```

dmuelle replied:
here too, the results focus on config (how) instead of value (why)

scottkurz replied:
I get what you're saying, but I think if we back up you'll see why I didn't go into more detail.    Backing up, all our Liberty-related Spring Boot doc focuses on what I'm calling the optimized deployment.   In fact, we don't even call it an "optimized deployment"... we just kind of take for granted this is how you deploy Spring Boot to the point that we don't even mention the "standard WAR" deployment.  

So I think there's enough out there overall about the thinning, etc. advantages of the optimized deployment. I did link back to it.  I don't think this one blog post needs to make the case in all that much detail.. it's been our story for awhile.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+== Results
+
+To recap, a Spring Boot WAR can be deployed to Liberty like any other WAR, or it can be deployed with an optimized deployment using special liberty-maven-plugin configuration and the `springBoot-3.0` feature configured in server.xml.
```
```suggestion
To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can deploy it with an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature configured in server.xml.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+To recap, a Spring Boot WAR can be deployed to Liberty like any other WAR, or it can be deployed with an optimized deployment using special liberty-maven-plugin configuration and the `springBoot-3.0` feature configured in server.xml.
+Though much of the programming model is the same across the two cases, there are some differences including the bootstrap mechanism.
+
```
recap the value

scottkurz replied:
I tried to address this overall.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+
+== References
+* Clone the https://github.com/scottkurz/spring-liberty[repository] with the finished code for this blog.
+* doc:   https://openliberty.io/docs/latest/deploy-spring-boot.html   
```
```suggestion
* Docs: link:https://openliberty.io/docs/latest/deploy-spring-boot.html[Configure and Deploy Spring Boot applications to Open Liberty]   
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+== References
+* Clone the https://github.com/scottkurz/spring-liberty[repository] with the finished code for this blog.
+* doc:   https://openliberty.io/docs/latest/deploy-spring-boot.html   
+* instanton: https://openliberty.io/blog/2023/09/26/spring-boot-3-instant-on.html
```
instantOn isn't explained or mentioned in the post, I'd remove this unless you add some context

scottkurz replied:
Done

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+* Clone the https://github.com/scottkurz/spring-liberty[repository] with the finished code for this blog.
+* doc:   https://openliberty.io/docs/latest/deploy-spring-boot.html   
+* instanton: https://openliberty.io/blog/2023/09/26/spring-boot-3-instant-on.html
+* GH repo:  https://github.com/scottkurz/spring-liberty
```
```suggestion
* link:https://github.com/scottkurz/spring-liberty[spring-liberty Git repo](https://github.com/scottkurz/spring-liberty)
```

dmuelle replied:
what is this? might help to add a sentence to explain

scottkurz replied:
I guess this was a dup, so I combined with the earlier link.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+* doc:   https://openliberty.io/docs/latest/deploy-spring-boot.html   
+* instanton: https://openliberty.io/blog/2023/09/26/spring-boot-3-instant-on.html
+* GH repo:  https://github.com/scottkurz/spring-liberty
+* Spring Boot Open Liberty guide:  https://openliberty.io/guides/spring-boot.html
```
```suggestion
* Guide: link:https://openliberty.io/guides/spring-boot.html[Containerizing, packaging, and running a Spring Boot application]  
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,232 @@
+* GH repo:  https://github.com/scottkurz/spring-liberty
+* Spring Boot Open Liberty guide:  https://openliberty.io/guides/spring-boot.html
+
+* TODO - way to note support for minor references
```
remove this, pending input from chuck or harry?

scottkurz replied:
done

dmuelle reviewed (changes requested):
ID review

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+:url-about: /
+//Blank line here is necessary before starting the body of the post.
+
+As of version 23.0.0.9, Liberty offers support for Spring Boot 3.1, allowing developers to take advantage of the Spring and Spring Boot libraries while providing a range of https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance] enhancements specifically engineered to provide an industry-leading, first-class platform for Spring Boot applications.
```
```suggestion
As of version 23.0.0.9, Liberty offers support for Spring Boot 3.1, including a range of https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance] enhancements that provide an industry-leading, first-class platform for Spring Boot applications.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+As of version 23.0.0.9, Liberty offers support for Spring Boot 3.1, allowing developers to take advantage of the Spring and Spring Boot libraries while providing a range of https://openliberty.io/blog/2022/10/17/memory-footprint-throughput-update.html[performance] enhancements specifically engineered to provide an industry-leading, first-class platform for Spring Boot applications.
+
+When packaging a Spring Boot app in the WAR format, developers and administrators get to choose from two Liberty deployment options, each with a different set of advantages.
```
```suggestion
When you package a Spring Boot app in the WAR format, you can choose from two Liberty deployment options, each with a different set of advantages.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+When packaging a Spring Boot app in the WAR format, developers and administrators get to choose from two Liberty deployment options, each with a different set of advantages.
+
+A Spring Boot WAR can simply be deployed like any standard Jakarta EE WAR using the same developer tooling and server configuration, and providing a common DevOps workflow as with your other Liberty WAR deployments.
```
```suggestion
You can deploy a Spring Boot WAR  like any standard Jakarta EE WAR by using the same developer tooling and server configuration, and providing a common DevOps workflow with your other Liberty WAR deployments.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+A Spring Boot WAR can simply be deployed like any standard Jakarta EE WAR using the same developer tooling and server configuration, and providing a common DevOps workflow as with your other Liberty WAR deployments.
+
+In addition, with Liberty, you can use a special https://openliberty.io/docs/latest/deploy-spring-boot.html[optimized deployment for Spring Boot applications]. The optimization adds some configuration options and provides advantages for efficiently building Docker container images by using some extra `pom.xml` and `server.xml` file configuration.
```
```suggestion
Alternatively, you can use a special https://openliberty.io/docs/latest/deploy-spring-boot.html[optimized Liberty deployment for Spring Boot applications]. This optimization adds some configuration options and provides advantages to efficiently build Docker container images.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+In this blog post, we'll first create a simple app by using the link:https://start.spring.io/[Spring Initializr], and then deploy it as a WAR to Liberty without any modifications, using the same deployment as for any WAR.
+
+We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and a simple `server.xml` configuration with the `springBoot-3.0` feature.   We can then build an efficiently-layered Docker image.
```
```suggestion
We then show how to optimize the Spring Boot deployment by adding the Liberty Maven Plugin (`liberty-maven-plugin`) and a simple `server.xml` configuration with the `springBoot-3.0` feature. We can then build an efficiently-layered Docker image.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+
+== Prerequisites
+- The sample requires Java 21 (though it could be easily reworked using Java 17)
```
```suggestion
- The sample requires Java 21 (though you could be easily rework it to use Java 17)
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+== Prerequisites
+- The sample requires Java 21 (though it could be easily reworked using Java 17)
+
+== Deploy as standard WAR
```
```suggestion
== Option 1: Deploy your Spring Boot app as standard WAR file
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+The Spring portion of the app is bootstrapped by the web container by using the `SpringBootServletInitializer` extension generated into your application by the **Spring Initializr** and from there you can use the Spring and Spring Boot APIs.
+
+== Use optimized Liberty Spring Boot deployment
```
```suggestion
== Option 2: Use the optimized Liberty Spring Boot deployment
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+== Deploy as standard WAR
+
+1. Create a project from your browser using link:https://start.spring.io/#!type=maven-project&language=java&platformVersion=3.2.5&packaging=war&jvmVersion=21&groupId=demo&artifactId=spring-liberty&name=spring-liberty&description=Demo%20project%20for%20Spring%20Boot&packageName=demo&dependencies=web[this Spring Initializr configuration].
```
I'm not going to suggest the edit from PR review mode, but I think this 1st step should be part of the prereqs- it's the same step for either procedure so better to add as a prereq than ask "optimized deployers" to go back to the option they already skipped over

scottkurz replied:
OK, I  see why you are saying this, which means the flow isn't as clear as I'd thought.   You are assuming that because these are two parallel options that someone following the blog instructions could do either Option 1 OR Option 2, and therefore could start right at Option 2.

If I did that though I'd have to move the generation of the "2. Add the following Spring web controller file" into the prereq... the Option 2. builds on the Option 1. instructions.

I think it works well enough walking you through both options... it's not like either one is too long in and of itself.

Do you think it's worth trying to make this all more clear somehow, (the article organization I'm explaining in this comment)

I wonder if I should go back and rename the headings since I've since changed them to Option 1... Option 2.

dmuelle replied:
ok- I did think this was an either-or scenario. If it's not, we shouldn't use "option..." in the section headings. Also need to tweak the intro a bit. Is this the intended flow?

1. Here's how you deploy a spring app like any other WAR
2. If you want to optimize that deployment, take what you did in (1) and add some more config

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+1. Create a project from your browser using link:https://start.spring.io/#!type=maven-project&language=java&platformVersion=3.2.5&packaging=war&jvmVersion=21&groupId=demo&artifactId=spring-liberty&name=spring-liberty&description=Demo%20project%20for%20Spring%20Boot&packageName=demo&dependencies=web[this Spring Initializr configuration].
++
+2. Add the following Spring web controller file (use the `vi` command here, or your preferred editor) :
```
```suggestion
2. Add the following Spring web controller file (use the `vi` command here, or your preferred editor).
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+== Use optimized Liberty Spring Boot deployment
+
+This example uses the same Spring app that is created in step 1 of the previous example. By adding some extra configuration in the `pom.xml` and `server.xml`, you can build an efficiently-layered Docker image and also parameterize the app with application arguments.
```
if you add the initializr to the prepreqs, update this section accordingly

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+This example uses the same Spring app that is created in step 1 of the previous example. By adding some extra configuration in the `pom.xml` and `server.xml`, you can build an efficiently-layered Docker image and also parameterize the app with application arguments.
+
+1. Add liberty-maven-plugin with Spring Boot deployment config to the `pom.xml` file:
```
```suggestion
1. Add the following `liberty-maven-plugin` with Spring Boot deployment config to the `pom.xml` file.
```

scottkurz replied:
That suggestion seemed a bit awkward to me.   While I'm not crazy about my next revision, I think the user will be clear enough what to do here, which is to copy and paste the XML snippet shown into the pom.xml.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+    </plugin>
+----
++
+2. Create your Liberty `server.xml` file by copying from the Spring Boot template
```
```suggestion
2. Create your Liberty `server.xml` file by copying from the Spring Boot template.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+cp target/liberty/wlp/usr/servers/defaultServer/server.xml src/main/liberty/config/server.xml
+----
++
+3. Add application arguments to your Spring Boot application configuration in the Liberty `server.xml` file
```
```suggestion
3. Add application arguments to your Spring Boot application configuration in the Liberty `server.xml` file.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+ </springBootApplication>
+----
++
+4. Add `println` to echo arguments in the output (use the `vi` command here, or your preferred editor) :
```
```suggestion
4. Add `println` to echo arguments in the output (use the `vi` command here, or your preferred editor).
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+    }
+----
++
+5. Run
```
```suggestion
5. Run
```

Run what? The app? the deployment?

scottkurz replied:
Done

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
++
+(Note you can use the `full-java17-openj9-ubi` tag to build the equivalent Java 17 image.)
++
+2. Build then run the image
```
```suggestion
2. Build then run the image.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+== Results
+
+To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can deploy it with an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature configured in server.xml.
```
```suggestion
To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can configure an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature in the `server.xml` file.
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+== Results
+
+To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can deploy it with an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature configured in server.xml.
+Though much of the programming model is the same across the two cases, there are some differences including the bootstrap mechanism.
```
```suggestion
Though much of the programming model is the same across the two cases, there are some differences, including the bootstrap mechanism and the ability to create more efficient Docker layers with the optimized deployment.
```

dmuelle reviewed (changes requested):
updated ID review

scottkurz commented:
> updated ID review

@dmuelle  Thanks.

To summarize my response... 
1. Here are most of your suggestions taken plus 1-2 that I did manually following your suggestion https://github.com/OpenLiberty/blogs/pull/3740
2. See my comment https://github.com/OpenLiberty/blogs/pull/3655#discussion_r1583799599  about how Option 2 depends on Option 1.

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+== Prerequisites
+- The sample requires Java 21 (though you could be easily rework it to use Java 17)
+
+== Option 1: Deploy your Spring Boot app as standard WAR file
```
```suggestion
== Deploy your Spring Boot app as standard WAR file
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+== Option 2: Use the optimized Liberty Spring Boot deployment
+
+This example uses the same Spring app that is created in step 1 of the previous example. By adding some extra configuration in the `pom.xml` and `server.xml`, you can build an efficiently-layered Docker image and also parameterize the app with application arguments.
```
```suggestion
This example uses the same Spring Initializr app that you created in the previous section. By adding some extra configuration in the `pom.xml` and `server.xml` files, you can build an efficiently-layered Docker image and also parameterize the app with application arguments.
```

scottkurz replied:
left alone after discussing

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,238 @@
+
+The Spring portion of the app is bootstrapped by the web container by using the `SpringBootServletInitializer` extension generated into your application by the **Spring Initializr** and from there you can use the Spring and Spring Boot APIs.
+
+== Option 2: Use the optimized Liberty Spring Boot deployment
```
```suggestion
== Optimize the deployment with the Liberty Spring Boot feature
```

dmuelle reviewed (changes requested):
updated my suggestions so it doesn't read like an either-or scenario

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -180,7 +180,7 @@ The server is configured with the https://openliberty.io/docs/latest/reference/f
 If you're wondering, yes, while a Spring Boot WAR must be repackaged as an executable WAR to use the optimized deployment, that executable WAR could still be deployed as a standard WAR.
 
-=== Build container image with efficient layering
+=== Part 3: Build container image with efficient layering
```
```suggestion
== Part 3: Build a container image with efficient layering
```

dmuelle commented on posts/2024-05-01-spring-boot-3.adoc:
```diff
@@ -0,0 +1,239 @@
+To recap, you can deploy a Spring Boot WAR to Liberty like any other WAR, or you can configure an optimized deployment using special `liberty-maven-plugin` configuration and the `springBoot-3.0` feature in the `server.xml` file.
+Though much of the programming model is the same across the two cases, there are some differences, including the bootstrap mechanism and the ability to create more efficient Docker layers with the optimized deployment.
+
+(**Note:** the flow through Parts 1 and 2 of this blog post are a bit contrived to show the two deployment options, and your real development workflow would probably look a bit different. When using a traditional WAR deployment you would likely maintain your own `server.xml` file rather than continuing to use the template configuration. When using the optimized Liberty deployment there's no need  ...
```
```suggestion
**Note:** The flow through Parts 1 and 2 of this blog post is a bit contrived to show the two deployment options, and your real development workflow would probably look a bit different. When using a traditional WAR deployment, you would likely maintain your own `server.xml` file rather than continuing to use the template configuration. When using the optimized Liberty deployment, there's no need to do a traditional deployment first.
```

dmuelle reviewed (approved):
LGTM- thanks!
